bindings/python: add 'flux_job_list_inactive' Python binding

cmoussa1 opened the pull request:
As described in #2776, there is no Python binding for `flux job list-inactive`. This command is planned to be used by the Bank/Accounting interface for Flux, which for the time being is being written in Python. 

This PR proposes the following changes:

- adds a Python binding for `flux job list-inactive` so it can be used in an RPC to get job record data.
- renames the **timestamp** option to **since** to more accurately describe what the option is doing. 
- changes the default value of **max_entries** in `job_list()` to **1000** to match the default value in `job_list_inactive()` (discussed in the comments thread in #2776).

cmoussa1 commented:
The **[WIP]** tag is there because I'm unsure how to integrate the new unit test file, **t0013-job-list-inactive.py** to be run when running `make check`.

chu11 commented on src/bindings/python/flux/job.py:
```diff
@@ -260,7 +260,7 @@ def job_list(flux_handle, max_entries=0, attrs=[], userid=os.geteuid(), states=0
 def job_list_inactive(flux_handle, since, max_entries=1000, attrs=[]):
     payload = {
-        "timestamp": since,
+        "since": since,
```
for commit message for this change, I might instead say "Rename the 'timestamp' payload field to 'since' in `job-info.list-inactive`, adjust appropriately in libjob/job.[ch] and rename --timestamp to --since in cmd/flux-job."  Or something like that.

chu11 commented:
> The [WIP] tag is there because I'm unsure how to integrate the new unit test file, t0013-job-list-inactive.py to be run when running make check.

This can be added in `t/Makefile.am`.

chu11 commented on t/python/t0013-job-list-inactive.py:
```diff
@@ -0,0 +1,263 @@
+#!/usr/bin/env python
```
for this commit message, perhaps say `t/python: Add new tests ...`

grondo commented on src/bindings/python/flux/job.py:
```diff
@@ -258,6 +258,11 @@ def job_list(flux_handle, max_entries=0, attrs=[], userid=os.geteuid(), states=0
     return JobListRPC(flux_handle, "job-info.list", payload)
 
 
+def job_list_inactive(flux_handle, since, max_entries=1000, attrs=[]):
```
Would it make sense to have a default of `since=0.`? That way `job_list_inactive(flux_handle)` returns the 1000 most recently inactive jobs. It is only a minor convenience, so not a big deal...

grondo commented on src/cmd/flux-job.c:
```diff
@@ -93,8 +93,8 @@ static struct optparse_option list_inactive_opts[] =  {
-    { .name = "timestamp", .key = 't', .has_arg = 1, .arginfo = "T",
-      .usage = "Limit output to jobs newer than timestamp",
+    { .name = "since", .key = 's', .has_arg = 1, .arginfo = "T",
+      .usage = "Limit output to jobs newer than since",
```
Suggested change for help string: `Limit output to jobs that entered the inactive state since timestamp T`

grondo commented on src/common/libjob/job.h:
```diff
@@ -107,11 +107,11 @@ flux_future_t *flux_job_list (flux_t *h,
 
 /* Similar to flux_job_list(), but retrieve inactive jobs newer
- * than a timestamp.
+ * than since.
```
I would just drop this change because the old description made slightly more sense. Or, change to `...retrieve inactive jobs newer than the timestamp "since"`

grondo commented on src/modules/job-info/list.c:
```diff
@@ -180,17 +180,17 @@ void list_cb (flux_t *h, flux_msg_handler_t *mh,
+/* Create a JSON array of 'job' objects.  'since' limits entries
  * returned, only returning entries with 't_inactive' newer than the
- * timestamp.  Returns JSON object which the caller must free.  On
+ * since.  Returns JSON object which the caller must free.  On
```
Similar to above: `s/since/timestamp "since"/`

chu11 commented on t/python/t0013-job-list-inactive.py:
```diff
@@ -0,0 +1,263 @@
+        # submit a sleep 0 job
+        self.submitJob()
+
+        time.sleep(1)
```
I think you should be calling `job.wait()` b/c this could be racy.

grondo commented on t/python/t0013-job-list-inactive.py:
```diff
@@ -0,0 +1,281 @@
+        # submit a sleep 0 job
+        self.submitJob()
+
+        time.sleep(1)
```
If possible, avoid sleeping in tests because this will slow the testsuite down. It appears that the job was submitted with `waitable=True`, so try using `job.wait()` here to wait just until the job completes to move on to the `job_list_inactive()` call.

chu11 commented on t/python/t0013-job-list-inactive.py:
```diff
@@ -0,0 +1,263 @@
+            print("{}: {}".format("rpc", e.strerror), file=sys.stderr)
+            sys.exit(1)
+
+        self.assertEqual(len(jobs), 9)
```
I think I'm missing something, why would the returned list be length 9?  Shouldn't it be atleast 10 given the `range(10)` above?  Also with the test below this one.

chu11 commented on t/python/t0013-job-list-inactive.py:
```diff
@@ -0,0 +1,263 @@
+        )
+
+        try:
+            t_inactive = rpc_handle.get_jobs()
```
I think `t_inactive[0]["t_inactive"]` is a little confusing.  Similar below in other tests.  One idea, perhaps all calls to `job_list()` use `jobs` and all calls to `job_list_inactive()` use `jobs_inactive` or something like that.

chu11 commented on t/python/t0013-job-list-inactive.py:
```diff
@@ -0,0 +1,263 @@
+    # flux job list-inactive with the most recent timestamp should return len(0)
+    def test_05_most_recent_inactive(self):
+        for i in range(5):
+            jobid = self.submitJob()
```
should wait on `jobid` to ensure it completes before tests below?  Or perhaps wait could be added to `submitJob()` so that you don't have to call `job.wait()` all the time?   Similar below in other tests.

chu11 commented on t/python/t0013-job-list-inactive.py:
```diff
@@ -0,0 +1,263 @@
+            print("{}: {}".format("rpc", e.strerror), file=sys.stderr)
+            sys.exit(1)
+
+        self.assertEqual(jobs[0]["t_inactive"], t_inactive[0]["t_inactive"])
```
also assert that list is a length 1?

chu11 commented on src/cmd/flux-job.c:
```diff
@@ -93,8 +93,8 @@ static struct optparse_option list_inactive_opts[] =  {
-    { .name = "timestamp", .key = 't', .has_arg = 1, .arginfo = "T",
-      .usage = "Limit output to jobs newer than timestamp",
+    { .name = "since", .key = 's', .has_arg = 1, .arginfo = "T",
+      .usage = "Limit output to jobs newer than since",
```
just saw @grondo's comment, I think his is better :P

SteVwonder commented on t/python/t0013-job-list-inactive.py:
```diff
@@ -0,0 +1,281 @@
+            jobs = rpc_handle.get_jobs()
+        except EnvironmentError as e:
+            print("{}: {}".format("rpc", e.strerror), file=sys.stderr)
+            sys.exit(1)
```
Nit (feel free to take or leave it): maybe move these five lines of code into a function/method (like you did with `submitJob`) to reduce the amount of duplicated code in the tests.

grondo commented:
> Although we aren't the most consistent with how we do titles in commit messages, job-info: and job.py: are a little short. I tend to do modules/job-info for job-info and and python/job seems to be the most popular for job.py.

Yeah, we definitely aren't consistent, and there may be some confusion about the purpose of this "subsystem" (for lack of a better term) prefix. I opened flux-framework/flux-docs#28 to suggest we come up with some better rules for this prefix, or at least figure out all the reasons we require it.

cmoussa1 commented:
Thanks for all of your feedback @chu11 @grondo @SteVwonder ! I've made some changes addressing the suggestions you gave me and pushed those up. Let me know if I missed anything (first time solving merge conflicts too, so I may have gone haywire somewhere without knowing 😬).

cmoussa1 commented on t/python/t0013-job-list-inactive.py:
```diff
@@ -0,0 +1,241 @@
+        flux.job.submit(self.fh, compute_jobreq, waitable=True)
+
+    @classmethod
+    def execRPC(self, rpc_handle):
```
Yeah, I agree with you - I'm not the biggest fan of that function name either. How about something like `getJobs()`? That could be general for both calling the `job_list()` and `job_list_inactive()` RPC's?

chu11 replied:
I think `getJobs()` is would be fine.

cmoussa1 commented on t/python/t0013-job-list-inactive.py:
```diff
@@ -0,0 +1,241 @@
+            )
+            jobs = self.execRPC(rpc_handle)
+            if len(jobs) == 1:
+                break
```
Yes, sounds like a good idea. Maybe call it something like `waitForConsistency()`? I hesitate to just call it `wait()` because that might be confusing to have to differentiate between the `wait()` RPC and our own "@classmethod" implementation.

chu11 replied:
`waitForConsistency()` sounds good to me!  Also, might be good to add some comments similar to the ones I have in `t2204-job-info` about how `job.wait()` syncs on the job manager, gotta wait for `job-info` eventual consistency, etc.

cmoussa1 commented:
Pushed some more changes per @chu11's suggestions. 

While trying to re-order the `since` patch commit, I kept running to some merge conflicts that I was unsure how to handle, so I avoided reordering them for the time being. Maybe we could walk through that together if you wouldn't mind @chu11 ?

grondo commented on src/cmd/flux-job.c:
```diff
@@ -93,8 +93,9 @@ static struct optparse_option list_inactive_opts[] =  {
-      .usage = "Limit output to jobs newer than timestamp",
+    { .name = "since", .key = 's', .has_arg = 1, .arginfo = "T",
+      .usage = "Limit output to jobs that entered the inactive state since \
+      timestamp T",
```
breaking the line in this way results in extra whitespace between `since` and `timestamp`:
```
Usage: flux-job list-inactive [OPTIONS]
List Inactive jobs
  -h, --help             Display this message.
  -c, --count=N          Limit output to N jobs
  -s, --since=T          Limit output to jobs that entered the inactive state
                         since       timestamp T
```

It is actually simple to break strings across multiple lines in C, just create another quoted string on the next line and it will be concatenated:
```C
    { .name = "since", .key = 's', .has_arg = 1, .arginfo = "T",
￼     .usage = "Limit output to jobs that entered the inactive state since"
￼              " timestamp T",
```

cmoussa1 replied:
TIL! Thanks for the clarification @grondo 👍 pushed that change up

chu11 commented on t/python/t0013-job-list-inactive.py:
```diff
@@ -0,0 +1,242 @@
+    # until all of the inactive jobs have reached INACTIVE state.
+    @classmethod
+    def waitForConsistency(self, jobs_list_length):
+        # wait for eventual consistency
```
nit: the function's comment is sufficient, can remove this old comment line :-)

grondo commented:
The PR title will go into the merge commit body when mergify auto-merges the PR, which is the only reason I mention this:

Two minor nits on the title:

 * bindings/python/job.py: is ok, but I think probably the `job.py` is redundant. You can just use `bindings/python` (or even just `python:` IMO)
 * The title is slightly misleading. The binding is not for `flux job list-inactive` (a command) but for `flux_job_list_inactive()` (a libflux API function)

cmoussa1 commented:
Thanks for your feedback/suggestions @grondo @chu11! I've pushed some more changes per your advice.

chu11 commented:
LGTM!  I've approved the PR.  Set the label merge-when-passing if you think it's all ready on your end.
